semigroups/semiring.py

```diff
@@ -1,81 +1,999 @@
+        44
+        >>> TropicalMinPlusSemiring(10).threshold()
+        10
+    '''
 
-    def __init___(self):
-        pass
+    def __init__(self, threshold):
+        if not isinstance(threshold, int):
+            raise TypeError
+
+        if threshold < 0:
+            raise ValueError
+
+        self._threshold = threshold
 
     def plus(self, x, y):
+        '''
+        A function to find the maximum of two elements of a tropical min plus
+        semiring, since this is the additive operation of a tropical min plus
+        semiring.
+
+        Args:
+            x (int or float):    One of the elements to be added.
+            y (int or float):    The other of the elements to be added.
+
+        Returns:
+            int or float:   The minimum of x and y.
+
+        Raises:
+            TypeError:  If x and y are not both ints or plus infinity.
+            ValueError: If either x or y is negative and not plus infinity, or
+                        if x or y is greater than the threshold.
+
+        Examples:
+            >>> from semigroups import TropicalMinPlusSemiring
+            >>> TropicalMinPlusSemiring(7).plus(float('inf'), 3)
+            3
+        '''
+
         if not ((isinstance(x, int) or x == float('inf'))
                 and (isinstance(y, int) or y == float('inf'))):
             raise TypeError
+
+        if x < 0 or y < 0:
+            raise ValueError
+
+        if ((x > self._threshold and x != float('inf')) or
+                (y > self._threshold and y != float('inf'))):
+            raise ValueError
+
         return min(x, y)
 
     def prod(self, x, y):
+        '''
+        A function to find the integer sum of two elements of a tropical min
+        plus semiring, since this is the multiplicative operation of a tropical
+        min plus semiring. If the integer sum is greater than :math:`t`, then
+        the result is :math:`t`.
+
+        Args:
+            x (int or float):    One of the elements to be multiplied.
+            y (int or float):    The other of the elements to be multplied.
+
+        Returns:
+            int or float:    x + y
+
+        Raises:
+            TypeError:  If x and y are not both ints or plus infinity.
+            ValueError: If either x or y is negative and not plus infinity, or
+                        if x or y is greater than the threshold.
+
+        Examples:
+            >>> from semigroups import TropicalMinPlusSemiring
+            >>> TropicalMinPlusSemiring(7).prod(float('inf'), 3)
+            inf
+        '''
+
         if not ((isinstance(x, int) or x == float('inf'))
                 and (isinstance(y, int) or y == float('inf'))):
             raise TypeError
-        return x + y
+        if x < 0 or y < 0:
+            raise ValueError
+        if ((x > self._threshold and x != float('inf')) or
+                (y > self._threshold and y != float('inf'))):
+            raise ValueError
+        if max(x, y) == float('inf'):
+            return x
```

Comment:
I think it should return min(x, y), as this checks if at least one of them equals infinity